kafka/kafka.go

```diff
@@ -312,26 +331,38 @@ func (ams *asyncMessageSource) passAcksToKafka(ctx context.Context, c *cluster.C
 					Expected: forAcking[0],
 				}
 			default:
-				if forAcking[0].offset != nil {
-					off := forAcking[0].offset
-					c.MarkPartitionOffset(off.topic, off.partition, off.offset, "")
-				} else {
-					c.MarkOffset(forAcking[0].cm, "")
-				}
+				forAcking[0].ack()
 				forAcking = forAcking[1:]
 			}
-		case err := <-c.Errors():
-			return err
-		case <-ctx.Done():
-			return c.Close()
 		}
 	}
 }
 
+func (ams *asyncMessageSource) ConsumeMessages(ctx context.Context, messages chan<- substrate.Message, acks <-chan substrate.Message) error {
+	rg, ctx := rungroup.New(ctx)
```

Comment:
Is it correct to use `rungroup` here? I may be wrong, but don't we want both functions to exit before cancelling the context?

Reply:
No, because running one without the other doesn't really make sense, if the ack passing terminates the message consumption will get blocked and the client won't be able to acknowledge the message. The message consumer can only terminate when the context is cancelled and when the message channel is closed (in that case we want to terminate fairly quickly as far as I understand, so we can't wait for the client to acknowledge messages).